### ISSUE TITLE
Running Forever in evaluation/inference mode

### ISSUE THREAD

**Issue by abdksyed** [PaddlePaddle/PaddleOCR#3924]
Hi,

I followed detection_en steps, to train for text localization on custom dataset using Pre Trained ResNet50 SAST weights.

```
Architecture : 
    Backbone : 
        layers : 50
        name : ResNet_SAST
    Head : 
        name : SASTHead
    Neck : 
        name : SASTFPN
        with_cab : True
    Transform : None
    algorithm : SAST
    model_type : det
Eval : 
    dataset : 
        data_dir : ../Forms_ICDAR_upd/
        label_file_list : ['../Forms_ICDAR_upd/train_label.txt']
        name : SimpleDataSet
        transforms : 
            DecodeImage : 
                channel_first : False
                img_mode : BGR
            DetLabelEncode : None
            DetResizeForTest : 
                resize_long : 1536
            NormalizeImage : 
                mean : [0.485, 0.456, 0.406]
                order : hwc
                scale : 1./255.
                std : [0.229, 0.224, 0.225]
            ToCHWImage : None
            KeepKeys : 
                keep_keys : ['image', 'shape', 'polys', 'ignore_tags']
    loader : 
        batch_size_per_card : 1
        drop_last : False
        num_workers : 2
        shuffle : False
Global : 
    cal_metric_during_train : False
    checkpoints : None
    debug : False
    distributed : False
    epoch_num : 50
    eval_batch_step : [600, 1000]
    infer_img : /content/Forms_ICDAR_upd/images/image.jpg
    log_smooth_window : 20
    pretrained_model : /content/PaddleOCR/output/sast_r50_vd_ic15/iter_epoch_50
    print_batch_step : 2
    save_epoch_step : 10
    save_inference_dir : None
    save_model_dir : ./output/sast_r50_vd_ic15/
    save_res_path : ./output/sast_r50_vd_ic15/predicts_sast.txt
    use_gpu : True
    use_visualdl : False
Loss : 
    name : SASTLoss
Metric : 
    main_indicator : hmean
    name : DetMetric
Optimizer : 
    beta1 : 0.9
    beta2 : 0.999
    lr : 
        learning_rate : 0.001
    name : Adam
    regularizer : 
        factor : 0
        name : L2
PostProcess : 
    expand_scale : 1.0
    name : SASTPostProcess
    nms_thresh : 0.2
    sample_pts_num : 2
    score_thresh : 0.5
    shrink_ratio_of_width : 0.3
Train : 
    dataset : 
        data_dir : ../Forms_ICDAR_upd/
        label_file_list : ['../Forms_ICDAR_upd/train_label.txt']
        name : SimpleDataSet
        ratio_list : [1.0]
        transforms : 
            DecodeImage : 
                channel_first : False
                img_mode : BGR
            DetLabelEncode : None
            SASTProcessTrain : 
                image_shape : [512, 512]
                max_text_size : 512
                min_crop_side_ratio : 0.3
                min_crop_size : 24
                min_text_size : 4
            KeepKeys : 
                keep_keys : ['image', 'score_map', 'border_map', 'training_mask', 'tvo_map', 'tco_map']
    loader : 
        batch_size_per_card : 4
        drop_last : False
        num_workers : 4
        shuffle : True
```

I converted my dataset to ICDAR 2015 format as mentioned in the document, and trained it for 50 epochs. When ever I am trying to evaluate or infer the image, the script is running forever. I tried to see where it's stopping, looks likt in the `infer_det` file, after model output, the code is stuck at `post_result = post_process_class(preds, shape_list)` line98.

**Comment by abdksyed** [comment]
@MissPenguin 
Any inputs will be highly helpful!

**Comment by MissPenguin** [comment]
What do you mean by “running forever”？ running slow or totally stuck？ Maybe you can add some log to check if the code is stuck in NMS stage.

As we mentioned in Python Inference tutorial: post-processing locality aware NMS has two versions: Python and C++. The speed of C++ version is obviously faster than that of Python version. Due to the compilation version problem of NMS of C++ version, C++ version NMS will be called only in Python 3.5 environment, and python version NMS will be called in other cases.

**Comment by abdksyed** [comment]
It's stuck.

I added logging information after every line to see where it's getting stuck,

```python
preds = model(images)
# preds is coming. It's getting stuck at post_process_class
post_result = post_process_class(preds, shape_list)
```

**Comment by abdksyed** [comment]
I tried to go down each function to see where it's stuck,

And after going to several files, I have arrived at `standard_nms` function in the `locality_aware_nums` file.
Here it's doing a while loop on order. I see that my order.size is 34120.
And after every iteration(which is taking few seconds), the order size is reducing by 8. Hence the long time.

Is the order.size value of 34120 normal??

**Comment by MissPenguin** [comment]
I found that you set the parameter `eval_batch_step` as [600, 1000], which means the evaluation step was triggered after only 600 iterations. In the early stage of training, the number of candidate bounding boxes is usually large, like 34120 in this case, which would make the NMS operation very time-consuming. So here are some tips you can try:
1. set `eval_batch_step` a larger value, like [5000, 2000]
2. use Python3.5, then the c++ version NMS function which is much faster will be called. [code](https://github.com/PaddlePaddle/PaddleOCR/blob/release/2.3/ppocr/postprocess/sast_postprocess.py#L145)

**Comment by abdksyed** [comment]
I trained for 50 epochs, which was around 600 iterations. I kept eval 600 is because to avoid eval while training, as eval during training was getting stuck.

Also, I see in my training log that the hmean after training for 50 epochs is 0. The loss is reducing, when training started it's 1.47, and by 50 epochs it has reduced to 0.92. But why at the end the hmean is 0?
```
[2021/09/06 17:24:23] root INFO: save model in ./output/sast_r50_vd_ic15/latest
[2021/09/06 17:24:23] root INFO: Initialize indexs of datasets:['../Forms_ICDAR_upd/train_label.txt']
[2021/09/06 17:24:31] root INFO: epoch: [49/50], iter: 576, lr: 0.001000, loss: 0.930596, score_loss: 0.791226, border_loss: 0.004178, tvo_loss: 0.088231, tco_loss: 0.004339, reader_cost: 2.76448 s, batch_cost: 3.52633 s, samples: 4, ips: 0.56716
[2021/09/06 17:24:33] root INFO: epoch: [49/50], iter: 578, lr: 0.001000, loss: 0.932700, score_loss: 0.794443, border_loss: 0.004122, tvo_loss: 0.088103, tco_loss: 0.004257, reader_cost: 0.00517 s, batch_cost: 1.45656 s, samples: 8, ips: 2.74620
[2021/09/06 17:24:38] root INFO: epoch: [49/50], iter: 580, lr: 0.001000, loss: 0.932031, score_loss: 0.795143, border_loss: 0.004122, tvo_loss: 0.088103, tco_loss: 0.004146, reader_cost: 0.88110 s, batch_cost: 2.34916 s, samples: 8, ips: 1.70274
[2021/09/06 17:24:41] root INFO: epoch: [49/50], iter: 582, lr: 0.001000, loss: 0.933616, score_loss: 0.795143, border_loss: 0.004122, tvo_loss: 0.088813, tco_loss: 0.004146, reader_cost: 0.00015 s, batch_cost: 1.44885 s, samples: 8, ips: 2.76080
[2021/09/06 17:24:44] root INFO: epoch: [49/50], iter: 584, lr: 0.001000, loss: 0.933616, score_loss: 0.791232, border_loss: 0.003841, tvo_loss: 0.089394, tco_loss: 0.004146, reader_cost: 0.00341 s, batch_cost: 1.38300 s, samples: 8, ips: 2.89226
[2021/09/06 17:24:47] root INFO: epoch: [49/50], iter: 586, lr: 0.001000, loss: 0.932031, score_loss: 0.791232, border_loss: 0.003629, tvo_loss: 0.088813, tco_loss: 0.003966, reader_cost: 0.00014 s, batch_cost: 1.37141 s, samples: 8, ips: 2.91671
[2021/09/06 17:24:48] root INFO: epoch: [49/50], iter: 587, lr: 0.001000, loss: 0.932200, score_loss: 0.787157, border_loss: 0.003364, tvo_loss: 0.089394, tco_loss: 0.003966, reader_cost: 0.00005 s, batch_cost: 0.68514 s, samples: 4, ips: 2.91911
[2021/09/06 17:24:54] root INFO: save model in ./output/sast_r50_vd_ic15/latest
[2021/09/06 17:24:54] root INFO: Initialize indexs of datasets:['../Forms_ICDAR_upd/train_label.txt']
[2021/09/06 17:25:04] root INFO: epoch: [50/50], iter: 588, lr: 0.001000, loss: 0.932200, score_loss: 0.786944, border_loss: 0.003045, tvo_loss: 0.088813, tco_loss: 0.003906, reader_cost: 4.25986 s, batch_cost: 5.01542 s, samples: 4, ips: 0.39877
[2021/09/06 17:25:07] root INFO: epoch: [50/50], iter: 590, lr: 0.001000, loss: 0.930328, score_loss: 0.783225, border_loss: 0.002777, tvo_loss: 0.089405, tco_loss: 0.003844, reader_cost: 0.00607 s, batch_cost: 1.49985 s, samples: 8, ips: 2.66693
[2021/09/06 17:25:10] root INFO: epoch: [50/50], iter: 592, lr: 0.001000, loss: 0.930328, score_loss: 0.783225, border_loss: 0.002708, tvo_loss: 0.089405, tco_loss: 0.003745, reader_cost: 0.05467 s, batch_cost: 1.51521 s, samples: 8, ips: 2.63989
[2021/09/06 17:25:13] root INFO: epoch: [50/50], iter: 594, lr: 0.001000, loss: 0.930328, score_loss: 0.783225, border_loss: 0.002922, tvo_loss: 0.089405, tco_loss: 0.003745, reader_cost: 0.00015 s, batch_cost: 1.42298 s, samples: 8, ips: 2.81100
[2021/09/06 17:25:15] root INFO: epoch: [50/50], iter: 596, lr: 0.001000, loss: 0.930328, score_loss: 0.792468, border_loss: 0.002922, tvo_loss: 0.089405, tco_loss: 0.003670, reader_cost: 0.00013 s, batch_cost: 1.37116 s, samples: 8, ips: 2.91725
[2021/09/06 17:25:18] root INFO: epoch: [50/50], iter: 598, lr: 0.001000, loss: 0.928862, score_loss: 0.787205, border_loss: 0.003189, tvo_loss: 0.089405, tco_loss: 0.003670, reader_cost: 0.00012 s, batch_cost: 1.37139 s, samples: 8, ips: 2.91676
[2021/09/06 17:25:20] root INFO: epoch: [50/50], iter: 599, lr: 0.001000, loss: 0.928862, score_loss: 0.783480, border_loss: 0.003393, tvo_loss: 0.090427, tco_loss: 0.003670, reader_cost: 0.00005 s, batch_cost: 0.68540 s, samples: 4, ips: 2.91801
[2021/09/06 17:25:25] root INFO: save model in ./output/sast_r50_vd_ic15/latest
[2021/09/06 17:25:31] root INFO: save model in ./output/sast_r50_vd_ic15/iter_epoch_50
[2021/09/06 17:25:31] root INFO: best metric, hmean: 0
```

**Comment by MissPenguin** [comment]
how many images in your training set? seems like the training set is too small.

have you tried the standard dataset for training? same problem?

**Comment by abdksyed** [comment]
Yeah, the images are very few, around 50. I am using the pre trained weights of ICDAR, and using my little data to kind of fine tune.
The images are scanned forms, which are having nearly 100 bounding boxes in each, so around 5000 bounding box in total with around 50 images.

**Comment by abdksyed** [comment]
I am expecting it to at least overfit to those 50 images, as I am try to train and test on same 50 images.
